torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
+
+    Example::
+        state = PowerSGDState(process_group=process_group, matrix_approximation_rank=1)
+        >>> ddp_model.register_comm_hook(state, powerSGD_hook)
+    """
+    process_group = state.process_group
+    group_to_use = process_group if process_group is not None else dist.group.WORLD
+    world_size = (
+        process_group.size() if process_group is not None else dist.get_world_size()
+    )
+
+    # The input tensor is a flattened 1D tensor.
+    input_tensor = bucket.get_tensors()[0]
+    device = input_tensor.device
+    dtype = input_tensor.dtype
+    # Unflatten the input tensor into per-parameter tensors, for layer-wise compression.
+    tensors = [
+        input_tensor[offset : offset + length].view(sizes)
+        for offset, length, sizes in zip(
+            bucket.get_offsets(), bucket.get_lengths(), bucket.get_sizes_list()
+        )
+    ]
+
+    # Step I: Handle rank-1 tensors.
+    # Allocate contiguous memory for rank-1 tensors to allreduce them without compression efficiently.
+    rank1_tensors = [tensor for tensor in tensors if tensor.ndimension() <= 1]
+    rank1_tensors_memory = (
+        torch.cat([tensor.view(-1) for tensor in rank1_tensors])
+        if rank1_tensors
+        else torch.tensor([], device=device)
+    )
+
+    # Step II: Handle high-rank tensors.
+    # Allocate contiguous memory for Ps and Qs to allreduce compressed high-rank tensors efficiently.
+    high_rank_tensors = [
+        tensor.view(tensor.shape[0], -1)
+        for tensor in tensors
+        if tensor.ndimension() > 1
+    ]
+    total_Ps_size = 0
+    ps_memory = None  # TODO(wayi): Store it in a dict of PowerState for warm-up.
+    total_Qs_size = 0
+    qs_memory = None  # TODO(wayi): Store it in a dict of PowerState for warm-up.
+    for tensor in high_rank_tensors:
+        n, m = tensor.shape
+        matrix_approximation_rank = min(n, m, state.matrix_approximation_rank)
+        total_Ps_size += n * matrix_approximation_rank
+        total_Qs_size += m * matrix_approximation_rank
+    ps_memory = torch.empty(total_Ps_size, device=device, dtype=dtype)
+    qs_memory = torch.empty(total_Qs_size, device=device, dtype=dtype)
+
+    # Create Ps and Qs that point to the allocated memory.
+    ps = []
+    qs = []
+    p_idx = 0
+    q_idx = 0
+    for tensor in high_rank_tensors:
+        n, m = tensor.shape
+        matrix_approximation_rank = min(n, m, state.matrix_approximation_rank)
+        ps.append(
+            ps_memory[p_idx : p_idx + n * matrix_approximation_rank].view(
+                n, matrix_approximation_rank
+            )
+        )
+        qs.append(
+            qs_memory[q_idx : q_idx + m * matrix_approximation_rank].view(
+                m, matrix_approximation_rank
+            )
+        )
+        p_idx += n * matrix_approximation_rank
+        q_idx += m * matrix_approximation_rank
+
+    # Initialize and then orthogonalize Qs.
+    with torch.random.fork_rng(devices=[]):
```

Comment:
Would it be reasonable to dedupe other use cases of this forking in grad compression to a helper function?

Reply:
Had the same feeling. However, in the first implementation, it has a loop in it: `for q in qs:`, between setting the manual seed and filling random values. It can be a bit tricky. Let me try to do it in a separate refactoring PR.